syntax = "proto3";

package example;


service Branch {
  rpc Query (QueryRequest) returns (QueryResponse);
  rpc Withdraw (WithdrawRequest) returns (Response);
  rpc Deposit (DepositRequest) returns (Response);
  rpc MsgDelivery (MsgDeliveryRequest) returns (MsgDeliveryResponse);
  rpc Propagate_Withdraw (PropagateWithdrawRequest) returns (Response);
  rpc Propagate_Deposit (PropagateDepositRequest) returns (Response);
}


message MsgDeliveryResponse {
    int32 branch_id = 5;
    string interface = 1;
    string result = 2;
    double balance = 3;
}
message MsgDeliveryRequest {
  int32 id = 1;
  string interface = 2;
  int32 money = 3;
  int32 branch = 4;
  repeated int32 writeset = 5;
}

message QueryRequest {
  int32 event_id = 1;
}

message QueryResponse {
  int32 balance = 1;
  string message = 2;
}

message WithdrawRequest {
  int32 event_id = 1;
  int32 amount = 2;
}

message DepositRequest {
  int32 event_id = 1;
  int32 amount = 2;
}

message Response {
  string message = 1;
}

message PropagateWithdrawRequest {
  int32 event_id = 1;
  int32 amount = 2;
  int32 branch =3;
}

message PropagateDepositRequest {
  int32 event_id = 1;
  int32 amount = 2;
  int32 branch =3;
}

message Request {
  string event_type = 1;
}
